File: Data Structures & Algorithms/two-integer-sum/submission-0.py
class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        
        tempDict = {}

        for index, num in enumerate(nums):
            tempDict[num] = index
        
        for index, num in enumerate(nums):
            keyNum = target - num
            if keyNum in tempDict and index != tempDict[keyNum]:
                return [index, tempDict[keyNum]]

        return []

File: Data Structures & Algorithms/combination-target-sum/submission-0.cpp
class Solution {
public:
    vector<vector<int>> ret;
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<int> subset;
        dfs(nums, subset, target, 0);
        return ret;
    }

private:
    void dfs(vector<int>& nums, vector<int>& subset, int target, int index) {
        if (target == 0) {
            ret.push_back(subset);
            return;
        }

        if (target < 0 || index >= nums.size()) {
            return;
        }   

        subset.push_back(nums[index]);
        dfs(nums, subset, target - nums[index], index);

        subset.pop_back();
        dfs(nums, subset, target, index + 1);
        
        return;
    }
};
